x/concentrated-liquidity/invariant_test.go

```diff
@@ -0,0 +1,163 @@
+//
+// Returns:
+// * All positions across all pools
+// * Total pool assets across all pools
+// * Total pool spread rewards across all pools
+// * Total pool incentives across all pools
+func (s *KeeperTestSuite) getAllPositionsAndPoolBalances(ctx sdk.Context) ([]model.Position, sdk.Coins, sdk.Coins, sdk.Coins) {
+	// Get total spread rewards distributed to all pools
+	allPools, err := s.clk.GetPools(ctx)
+	totalPoolAssets, totalSpreadRewards, totalIncentives := sdk.NewCoins(), sdk.NewCoins(), sdk.NewCoins()
+
+	// Sum up pool balances across all pools
+	for _, pool := range allPools {
+		clPool, ok := pool.(types.ConcentratedPoolExtension)
+		s.Require().True(ok)
+		totalPoolAssets = totalPoolAssets.Add(s.App.BankKeeper.GetAllBalances(ctx, clPool.GetAddress())...)
+		totalSpreadRewards = totalSpreadRewards.Add(s.App.BankKeeper.GetAllBalances(ctx, clPool.GetSpreadRewardsAddress())...)
+		totalIncentives = totalIncentives.Add(s.App.BankKeeper.GetAllBalances(ctx, clPool.GetIncentivesAddress())...)
+	}
+
+	// Get all positions in state
+	allPoolPositions, err := s.clk.GetAllPositions(ctx)
+	s.Require().NoError(err)
+
+	return allPoolPositions, totalPoolAssets, totalSpreadRewards, totalIncentives
+}
+
+// assertTotalRewardsInvariant asserts two invariants on the current context:
+// 1. Claiming spread rewards and incentives for all positions in state yields the amount stored in pool reward addresses minus rounding errors
+// 2. Claiming spread rewards and incentives for all positions in state empties all pool reward addresses except for rounding errors
+//
+// This function operates on cached context to avoid persisting any changes to state.
+func (s *KeeperTestSuite) assertTotalRewardsInvariant() {
+	// Get all positions and total pool balances across all CL pools in state
+	allPositions, initialTotalPoolLiquidity, expectedTotalSpreadRewards, expectedTotalIncentives := s.getAllPositionsAndPoolBalances(s.Ctx)
+
+	// Switch to cached context to avoid persisting any changes to state
+	cachedCtx, _ := s.Ctx.CacheContext()
+
+	// Collect spread rewards for all positions and track output
+	totalCollectedSpread, totalCollectedIncentives := sdk.NewCoins(), sdk.NewCoins()
+	for _, position := range allPositions {
+		owner, err := sdk.AccAddressFromBech32(position.Address)
+		s.Require().NoError(err)
+
+		// Log initial position owner balance
+		initialBalance := s.App.BankKeeper.GetAllBalances(cachedCtx, owner)
+
+		// Collect spread rewards.
+		collectedSpread, err := s.clk.CollectSpreadRewards(cachedCtx, owner, position.PositionId)
+		s.Require().NoError(err)
+
+		// Collect incentives.
+		//
+		// Since we expect forfeited coins to go to other positions who have not yet claimed, we
+		// do not include them in the sum.
+		//
+		// Balancer full range incentives are also not factored in because they are claimed and sent to
+		// gauge immediately upon distribution.
```

Comment:
I just wanted to make clear some logic that recently was solidified in my bug fix PR:

Image three positions, all equal. Position 1 claims too early and forfeits 100 usdc. What this does __not__ do, is it does not give positions 2 and 3 claim to 50 usdc each. What it does do is it gives all three positions claim to 33.33 usdc each. I think this actually makes sense though, since forfeiting is just putting those rewards back into the accumulator, so that shouldn't stop position 1 from eventually being part of retrieving those incentives. They just share it with others rather than having it to themselves.

Just wanted to write this logic out here to make sure there is consensus / understanding with how this works.

Reply:
Discussed offline but need to converge on forfeiting logic. That being said, this helper simply asserts the invariant that `totalClaimedIncentives = poolIncentiveBalance`, which should hold regardless of how we handle forfeits.